packages/teleterm/src/services/tshd/createClient.ts

```diff
@@ -413,8 +415,37 @@ export default function createClient(
         });
       });
     },
-  };
 
+    transferFile(
+      options: api.FileTransferRequest.AsObject,
+      listeners: FileTransferListeners,
+      abortSignal?: types.TshAbortSignal
+    ): void {
+      const req = new api.FileTransferRequest()
+        .setLogin(options.login)
+        .setSource(options.source)
+        .setDestination(options.destination)
+        .setServerId(options.serverId)
+        .setClusterId(options.clusterId)
+        .setServerUri(options.serverUri)
+        .setDirection(options.direction);
+
+      const call = tshd.transferFile(req);
+      abortSignal.addEventListener(() => call.cancel());
+
+      call.on('data', (data: api.FileTransferProgress) =>
+        listeners.onProgress(data.getPercentage())
+      );
+      call.on('end', () => {
+        listeners.onComplete();
+        call.removeAllListeners();
+      });
+      call.on('error', error => {
+        listeners.onError(error);
+        call.removeAllListeners();
+      });
```

Comment:
I wonder if perhaps a better idea would be to make the tshd client return just the `call` object and then contain the rest of the logic within `FileTransferClient`.

Right now all `FileTransferClient` does is just pass some stuff further down – at this point one might wonder why we even bother creating this class at all.

See how I organized the stream stuff in [my initial draft PR for tshd-initiated communication](https://github.com/gravitational/webapps/pull/1176/files#diff-f59f2f399d4ba6b8395ee8cacb38f0d2ac18fb12050554e9249747f04efb2371). I'm talking about the files in `services/tshd`. Though here you probably don't want to call `FileTransferClient` from within tshd client, but rather as I suggested, tsdh client would return the call object and then `FileTransferClient` could operate on it.

File transfer doesn't need to follow it 1:1. I think the files you created so far are alright it's just a matter of moving the logic around a bit.

Reply:
Unfortunately, context bridge doesn't allow us to return `call` object, but I followed the structure from linked PR, so now we have a function `createFileTransferStream` that is returned from tshd client.

Also, as discussed in DM's, we decided not to pass `FileTransferListeners` as an argument, but instead we return it from the tshd/http calls. It makes the code easier to understand, especially `getDownloader`/`getUploader` parts (less arguments, one callback instead of a callback returning another callback).
I also added `createFileTransferEventsEmitter` utility function that makes working with the events/listeners more convenient.